src/Avalonia.Visuals/Rendering/ImmediateRenderer.cs

```diff
@@ -0,0 +1,278 @@
+﻿// Copyright (c) The Avalonia Project. All rights reserved.
+// Licensed under the MIT license. See licence.md file in the project root for full license information.
+
+using System;
+using Avalonia.Platform;
+using Avalonia.VisualTree;
+using System.Collections.Generic;
+using Avalonia.Media;
+using System.Linq;
+
+namespace Avalonia.Rendering
+{
+    /// <summary>
+    /// A renderer which renders the state of the visual tree without an intermediate scene graph
+    /// representation.
+    /// </summary>
+    /// <remarks>
+    /// The immediate renderer supports only clip-bound-based hit testing; a control's geometry is
+    /// not taken into account.
+    /// </remarks>
+    public class ImmediateRenderer : RendererBase, IRenderer, IVisualBrushRenderer
+    {
+        private readonly IVisual _root;
+        private readonly IRenderRoot _renderRoot;
+        private IRenderTarget _renderTarget;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ImmediateRenderer"/> class.
+        /// </summary>
+        /// <param name="root">The control to render.</param>
+        public ImmediateRenderer(IVisual root)
+        {
+            Contract.Requires<ArgumentNullException>(root != null);
+
+            _root = root;
+            _renderRoot = root as IRenderRoot;
+        }
+
+        /// <inheritdoc/>
+        public bool DrawFps { get; set; }
+
+        /// <inheritdoc/>
+        public bool DrawDirtyRects { get; set; }
+
+        /// <inheritdoc/>
+        public void Paint(Rect rect)
+        {
+            if (_renderTarget == null)
+            {
+                _renderTarget = ((IRenderRoot)_root).CreateRenderTarget();
+            }
+
+            try
+            {
+                using (var context = new DrawingContext(_renderTarget.CreateDrawingContext(this)))
+                {
+                    using (context.PushTransformContainer())
+                    {
+                        Render(context, _root, _root.Bounds);
+                    }
+
+                    if (DrawDirtyRects)
+                    {
+                        var color = (uint)new Random().Next(0xffffff) | 0x44000000;
+                        context.FillRectangle(
+                            new SolidColorBrush(color),
+                            rect);
+                    }
+
+                    if (DrawFps)
+                    {
+                        RenderFps(context.PlatformImpl, _root.Bounds, true);
+                    }
+                }
+            }
+            catch (RenderTargetCorruptedException ex)
+            {
+                Logging.Logger.Information("Renderer", this, "Render target was corrupted. Exception: {0}", ex);
+                _renderTarget.Dispose();
+                _renderTarget = null;
+            }
+        }
+
+        /// <inheritdoc/>
+        public void Resized(Size size)
```

Comment:
From what I could tell, windowing systems automatically invalidate the window on a resize so it should be unnecessary.